tuf/ngclient/_internal/metadata_bundle.py

```diff
@@ -0,0 +1,461 @@
+                < self.timestamp.signed.meta["snapshot.json"].version
+            ):
+                # TODO not sure about the correct exception here
+                raise exceptions.ReplayedMetadataError(
+                    "snapshot",
+                    new_timestamp.signed.meta["snapshot.json"].version,
+                    self.timestamp.signed.meta["snapshot.json"].version,
+                )
+
+        if new_timestamp.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New timestamp is expired")
+
+        self._bundle["timestamp"] = new_timestamp
+        logger.debug("Updated timestamp")
+
+    # TODO: remove pylint disable once the hash verification is in metadata.py
+    def update_snapshot(self, data: bytes):  # pylint: disable=too-many-branches
+        """Verifies and loads 'data' as new snapshot metadata.
+
+        Args:
+            data: unverified new snapshot metadata as bytes
+
+        Raises:
+            RepositoryError: Metadata failed to load or verify. The actual
+                error type and content will contain more details.
+        """
+
+        if self.timestamp is None:
+            raise RuntimeError("Cannot update snapshot before timestamp")
+        if self.targets is not None:
+            raise RuntimeError("Cannot update snapshot after targets")
+        logger.debug("Updating snapshot")
+
+        meta = self.timestamp.signed.meta["snapshot.json"]
+
+        # Verify against the hashes in timestamp, if any
+        hashes = meta.hashes or {}
+        for algo, stored_hash in hashes.items():
+            digest_object = sslib_hash.digest(algo)
+            digest_object.update(data)
+            observed_hash = digest_object.hexdigest()
+            if observed_hash != stored_hash:
+                # TODO: Error should derive from RepositoryError
+                raise exceptions.BadHashError(stored_hash, observed_hash)
+
+        try:
+            new_snapshot = Metadata.from_bytes(data)
+        except DeserializationError as e:
+            raise exceptions.RepositoryError("Failed to load snapshot") from e
+
+        if new_snapshot.signed.type != "snapshot":
+            raise exceptions.RepositoryError(
+                f"Expected 'snapshot', got '{new_snapshot.signed.type}'"
+            )
+
+        if not verify_with_threshold(self.root, "snapshot", new_snapshot):
+            raise exceptions.UnsignedMetadataError(
+                "New snapshot is not signed by root", new_snapshot.signed
+            )
+
+        if (
+            new_snapshot.signed.version
+            != self.timestamp.signed.meta["snapshot.json"].version
+        ):
+            raise exceptions.BadVersionNumberError(
+                f"Expected snapshot version"
+                f"{self.timestamp.signed.meta['snapshot.json'].version},"
+                f"got {new_snapshot.signed.version}"
+            )
+
+        if self.snapshot:
```

Comment:
these properties are shorthand for `self._bundle.get("snapshot", None)` (they exist for the top level metadata only):
I like the properties because there is no indexing by name, and top-level metadata needs to be accessed a lot, but it does create another layer of abstraction for the reader ...